File: app/src/main/java/br/com/pedrosilva/tecnonutri/data/repositories/ProfileRepositoryImpl.java
package br.com.pedrosilva.tecnonutri.data.repositories;

import android.content.Context;

import java.util.List;

import br.com.pedrosilva.tecnonutri.data.converters.FeedDataModelConverter;
import br.com.pedrosilva.tecnonutri.data.converters.ProfileDataModelConverter;
import br.com.pedrosilva.tecnonutri.data.entities.ProfileResponse;
import br.com.pedrosilva.tecnonutri.data.net.ApiServiceGenerator;
import br.com.pedrosilva.tecnonutri.data.net.services.ProfileService;
import br.com.pedrosilva.tecnonutri.domain.entities.FeedItem;
import br.com.pedrosilva.tecnonutri.domain.entities.Profile;
import br.com.pedrosilva.tecnonutri.domain.repositories.ProfileRepository;
import retrofit2.Response;

/**
 * Created by psilva on 3/18/17.
 */

public class ProfileRepositoryImpl implements ProfileRepository {
    private Context context;

    public ProfileRepositoryImpl(Context context) {
        this.context = context;
    }

    @Override
    public void get(int userId, ProfileCallback callback) {
        ProfileService profileService = ApiServiceGenerator.getService(ProfileService.class);
        try {
            Response<ProfileResponse> response = profileService.profile(userId).execute();
            if (response.isSuccessful()) {
                ProfileResponse profileResponse = response.body();
                Profile profile = ProfileDataModelConverter.convertToDomainModel(profileResponse.getProfile());
                List<FeedItem> feedItems = FeedDataModelConverter.convertListToDomainModel(profileResponse.getItems());
                callback.onSuccess(profile, feedItems, profileResponse.getPage(), profileResponse.getTimestamp());
            }
        } catch (Exception e) {
            callback.onError(e);
        }
    }

    @Override
    public void loadMorePosts(int userId, int page, int timestamp, ProfileCallback callback) {
        ProfileService profileService = ApiServiceGenerator.getService(ProfileService.class);
        try {
            Response<ProfileResponse> response = profileService.posts(userId, page, timestamp).execute();
            if (response.isSuccessful()) {
                ProfileResponse profileResponse = response.body();
                Profile profile = ProfileDataModelConverter.convertToDomainModel(profileResponse.getProfile());
                List<FeedItem> feedItems = FeedDataModelConverter.convertListToDomainModel(profileResponse.getItems());
                callback.onSuccess(profile, feedItems, profileResponse.getPage(), profileResponse.getTimestamp());
            }
        } catch (Exception e) {
            callback.onError(e);
        }
    }
}

File: app/src/main/kotlin/br/com/pedrosilva/tecnonutri/data/repositories/ProfileRepositoryImpl.kt
package br.com.pedrosilva.tecnonutri.data.repositories

import br.com.pedrosilva.tecnonutri.data.converters.asDTO
import br.com.pedrosilva.tecnonutri.data.entities.ProfileResponse
import br.com.pedrosilva.tecnonutri.data.ext.getBodyOrThrow
import br.com.pedrosilva.tecnonutri.data.net.ApiServiceGenerator
import br.com.pedrosilva.tecnonutri.data.net.services.ProfileService
import br.com.pedrosilva.tecnonutri.domain.repositories.ProfileRepository
import br.com.pedrosilva.tecnonutri.domain.repositories.ProfileRepository.ProfileCallback
import retrofit2.Call

class ProfileRepositoryImpl : ProfileRepository {

    private val apiService by lazy {
        ApiServiceGenerator.getService<ProfileService>()
    }

    override operator fun get(userId: Int, callback: ProfileCallback) {
        apiService.profile(userId).handleResponse(callback)
    }

    override fun loadMorePosts(userId: Int, page: Int, timestamp: Int, callback: ProfileCallback) {
        apiService.posts(userId, page, timestamp).handleResponse(callback)
    }

    private fun Call<ProfileResponse>.handleResponse(callback: ProfileCallback) {
        try {
            val response = getBodyOrThrow()
            val (page, profile, feed, timestamp) = response
            callback.onSuccess(profile.asDTO(), feed.asDTO(), page, timestamp)
        } catch (e: Exception) {
            callback.onError(e)
        }
    }
}